### ISSUE TITLE
[Enhancement] Provide code fixes/actions for warnings

### ISSUE THREAD

**Issue by dannoe** [StefanMaron/BusinessCentral.LinterCop#202]
I was curious how Microsoft code actions (e.g. fix "Implicit With") were implemented and if we could also extend them easily by just implementing a base class or interface. It turned out to be not that trivial, but I managed to do it anyway:

Basically every code action is just a class which implements the ``ICodeFixProvider`` interface. But they are loaded quite differently than the ``DiagnosticAnalyzer``. Microsoft is using [MEF](https://docs.microsoft.com/en-us/dotnet/framework/mef/) for loading them, and it is not extendable that trivial (at least I don't know a way). So I added a reference to the [Harmony](https://github.com/pardeike/Harmony) library to rewrite some IL at runtime, so the BusinessCentral.LinterCop assembly is added to the list of assemblies loaded by MEF. I also implemented a basic code fix for LC0001 (FlowFields should not be editable) as an example.
You can have a look at the implementation at https://github.com/dannoe/BusinessCentral.LinterCop/tree/add-code-actions

Would you accept a PR to add this into BusinessCentral.LinterCop or should I start my own repo/project for this?

**Comment by StefanMaron** [comment]
Oh please just throw in the PRs.
I did not do any code fixes yet, simply because I do not understand how to implement them

**Comment by StefanMaron** [comment]
maybe that helps?
![image](https://user-images.githubusercontent.com/24838311/197553347-f3433ec0-153a-4baf-984c-61ff10d6a178.png)

**Comment by StefanMaron** [comment]
I asked, they said no :D

To come back to the original question, I think we can go this way, but only if we can create this as a separate dll to not introduce a dependency to the diagnostics part of the linter.

What do you think?

**Comment by dannoe** [comment]
It looks like Harmony currently doesn't support patching a .NET 6 dll via [transpiler](https://harmony.pardeike.net/articles/patching-transpiler.html). At the time of writing this proof-of-concept, the default runtime of the AL Editor backend must have been .NET Framework 4.8.

I opened an issue on the harmony repo, maybe it's easy to implement. (see https://github.com/pardeike/Harmony/issues/503)
If not I am thinking about using the "normal" way of patching with harmony.

> I asked, they said no :D

Did they provide a reason?

> To come back to the original question, I think we can go this way, but only if we can create this as a separate dll to not introduce a dependency to the diagnostics part of the linter.
> 
> What do you think?

Sounds okay. Should this extra project be a part of this repository or have its own repository?

**Comment by StefanMaron** [comment]
The reason was that the only reason to open up the class and make this kind of changes would be to officially support extending it, and they dont want to do that.

Well, the code actions project would be 100% depending on the overwriting of that class, and I am not able to support that at all :D 
But I think I could add this the the linter vs code extension to handle the auto download if it would be a separate repo.

**Comment by dannoe** [comment]
The problem with harmony was the following: 
The default .NET runtime of the AL compiler is .NET 6 but the LinterCop is still compiled with .NET Framework 4.8. As I was using harmony via NuGet, it grabs the .NET FX 4.8 version of Harmony. As a result harmony tries to use methods from the .NET Framework which are non existent in .NET 6.

If I find the time to setup a new repository for the CodeFixes I will do so and comment on this issue.

**Comment by f4n0** [comment]
I think the only way is to add the vscode handler inside the VScode Lintercop extension. 
just because you need to extend the vscode.CodeActionProvider ts class and register your providers.

**Comment by Arthurvdv** [comment]
You're absolutely right, @dannoe. While the AL Language is evolving rapidly, it may take some time before we see _official_ support for custom code cops or code fix actions.

Currently, the custom code cops is kinda manageable, but I understand that introducing official support would slow down AL Language development, as it would need to ensure compatibility and avoid breaking changes.

For code fix/actions, having an initial step to enable custom development would be a great start. Even without official support, opening this up would help. As of now, it seems completely closed off. We could push this forward by adding more votes to [this BC Idea](https://experience.dynamics.com/ideas/idea/?ideaid=5629bee9-814e-ef11-b4ac-0022484d1652).

**Comment by Arthurvdv** [comment]
From AL version 15.0 and up, custom code actions are now supported 🥳

**Comment by dannoe** [comment]
@Arthurvdv Was this announced somewhere or did you find out by looking at the de-compilation? Just curious if there's a video I really need to watch :)

**Comment by Arthurvdv** [comment]
@dannoe, I am unsure if there is an official announcement, however, @MODUSCarstenScholling brought this to my attention here: https://github.com/StefanMaron/BusinessCentral.LinterCop/discussions/827.

In the [v0.35.0](https://github.com/StefanMaron/BusinessCentral.LinterCop/releases/tag/v0.35.0) release of LinterCop, the first code action is included 🥳
